### PR TITLE
YLP-1132 Implement GET users not authenticated and DELETE user in shoreline

### PR DESCRIPTION
**Pull request by yanngit**

### CODE REVIEW

File: user/api.go [hunk 1]
```diff
@@ -357,12 +358,20 @@ func (a *Api) GetUsers(res http.ResponseWriter, req *http.Request) {
 	} else if role := req.URL.Query().Get("role"); role != "" && !IsValidRole(role) {
 		a.sendError(res, http.StatusBadRequest, STATUS_INVALID_ROLE)
 
+	} else if authenticated := req.URL.Query().Get("authenticated"); authenticated != "" && authenticated != "false" && authenticated != "true" {
```
Comment (yvon-dblg) [inline user/api.go]: You could use the error return of strconv.ParseBool() to see if the value is valid ?

**Review by yvon-dblg — COMMENTED** [review]
The authenticated parameter is misleading.
See comment:
```golang
EmailVerified  bool                   `json:"emailVerified" bson:"authenticated"` //tag is name `authenticated` for historical reasons
```
We should start transitioning with a really explicit name.

File: user/api.go (hunk 1 above)
Comment (yvon-dblg) [reply user/api.go]: Missing update on the OpenAPI documentation

**Review by yvon-dblg — CHANGES_REQUESTED** [review]

File: user/api.go (hunk 1 above)
Comment (yanngit) [reply user/api.go]: done

File: clients/shoreline/shorelineMock.go [hunk 2]
```diff
@@ -66,6 +66,16 @@ func (client *ShorelineMockClient) GetUser(userID, token string) (*schema.UserDa
...
+	}, nil
+}
+
 func (client *ShorelineMockClient) UpdateUser(userID string, userUpdate schema.UserUpdate, token string) error {
 	return nil
 }
+
+func (client *ShorelineMockClient) DeleteUser(userID string, token string) error {
+	return nil
+}
```
Comment (yvon-dblg) [inline clients/shoreline/shorelineMock.go]: Missing new line at the end of file

File: user/api.go [hunk 3]
```diff
@@ -337,6 +338,7 @@ func (a *Api) GetStatus(res http.ResponseWriter, req *http.Request) {
 // @Produce  json
 // @Param role query string false "Role" Enums(clinic)
 // @Param id query string false "List of UserId separated by ,"
+// @Param emailVerified query string false "Filter users on emailVerified"
```
Comment (yvon-dblg) [inline user/api.go]:
```suggestion
// @Param emailVerified query boolean false "Filter users on emailVerified"
```

File: user/user.go [hunk 4]
```diff
@@ -168,6 +169,11 @@ func IsValidRole(role string) bool {
 	}
 }
 
+func IsValidBoolean(val string) bool {
+	_, err := strconv.ParseBool(val)
+	return  err == nil
```
Comment (yvon-dblg) [inline user/user.go]:
```suggestion
	return err == nil
```

File: user/api.go [hunk 5]
```diff
@@ -357,12 +359,20 @@ func (a *Api) GetUsers(res http.ResponseWriter, req *http.Request) {
 	} else if role := req.URL.Query().Get("role"); role != "" && !IsValidRole(role) {
 		a.sendError(res, http.StatusBadRequest, STATUS_INVALID_ROLE)
 
-	} else if userIds := strings.Split(req.URL.Query().Get("id"), ","); len(userIds[0]) > 0 && role != "" {
+	} else if emailVerified := req.URL.Query().Get("emailVerified"); emailVerified != "" && !IsValidBoolean(emailVerified) {
+		a.sendError(res, http.StatusBadRequest, STATUS_INVALID_EMAIL_VERIF_BOOL_PARAM)
+
+	} else if userIds := strings.Split(req.URL.Query().Get("id"), ","); len(req.URL.Query()) > 1 {
```
Comment (yvon-dblg) [inline user/api.go]: Missing comment to explain the `len(req.URL.Query()) > 1` is here

Comment (yvon-dblg) [reply user/api.go]: This code (not just your change) is not very clear sadly.

File: user/mongoStoreClient.go [hunk 6]
```diff
@@ -105,6 +105,11 @@ func (c *Client) FindUsersByRole(ctx context.Context, role string) (results []*U
 	return c.findUsers(ctx, bson.M{"roles": role}, noUserMessage)
 }
 
+func (c *Client) FindUsersByEmailVerified(ctx context.Context, emailVerified bool) ([]*User, error) {
+	noUserMessage := fmt.Sprintf("no users found: query: emailVerified: %v", emailVerified)
+	return c.findUsers(ctx, bson.M{"emailVerified": emailVerified}, noUserMessage)
```
Comment (yvon-dblg) [inline user/mongoStoreClient.go]: The mongodb field is still `authenticated` sadly.
```suggestion
	return c.findUsers(ctx, bson.M{"authenticated": emailVerified}, noUserMessage)
```

**Review by yvon-dblg — COMMENTED** [review]
Mostly not very important remarks,
just one potential problem on the mongodb search field.

**Review by yvon-dblg — COMMENTED** [review]
Missing changelog

**Review by yvon-dblg — APPROVED** [review]
Approved to not block the merged while I'm in holiday.

**Review by ydrisR — APPROVED** [review]
lgtm BUT the current delete logic is bizare... It requires a password to be provided in the payload, even if the request is comming with a server token and the password can be anything (there is no verification)
I approve the PR as this is not in the scope but we should probably rework that.

**Comment by yanngit** [comment]
@ydrisR  I changed a bit the API for server token, we're not checking the password anymore :)

File: user/api.go [hunk 7]
```diff
@@ -608,7 +621,7 @@ func (a *Api) GetUserInfo(res http.ResponseWriter, req *http.Request, vars map[s
 // @Accept  json
 // @Produce  json
 // @Param userid path int true "user id for server request, from token for personal request" optional
-// @Param password body string true "password"
+// @Param password body string true "password for personal request"
```
Comment (CedricPerotto) [inline user/api.go]: This parameter isn't required anymore 
```suggestion
// @Param password body string false "password mandatory for personal request"
```

File: user/api.go [hunk 8]
```diff
@@ -340,6 +341,7 @@ func (a *Api) GetStatus(res http.ResponseWriter, req *http.Request) {
 // @Produce  json
 // @Param role query string false "Role" Enums(clinic)
 // @Param id query string false "List of UserId separated by ,"
+// @Param emailVerified query boolean false "Filter users on emailVerified"
```
Comment (CedricPerotto) [inline user/api.go]: Out of scope just giving more context to OpenApi doc.

```suggestion
// @Param role query string false "Role. Exactly one query parameter is required in role, id, emailVerified" Enums(clinic)
// @Param id query string false "List of UserId separated by ,. Exactly one query parameter is required in role, id, emailVerified"
// @Param emailVerified query boolean false "Filter users on emailVerified. Exactly one query parameter is required in role, id, emailVerified"
```

File: user/api.go [hunk 9]
```diff
@@ -635,7 +648,7 @@ func (a *Api) DeleteUser(res http.ResponseWriter, req *http.Request, vars map[st
 
 	pw := getGivenDetail(req)["password"]
 
-	if id != "" && pw != "" {
+	if id != "" && (td.IsServer || pw != "") {
```
Comment (CedricPerotto) [inline user/api.go:651]: Looks like we're missing some tests on Server token + missing password.

Comment (yanngit) [reply user/api.go:651]: Test done

**Pull request merged**